### PR TITLE
MappingBuilder enable/disabled add and remove buttons

### PR DESCRIPTION
**Pull request by gildub**
Resolves https://github.com/konveyor/virt-ui/issues/42, https://github.com/konveyor/virt-ui/issues/96

### CODE REVIEW

File: src/app/Mappings/components/MappingBuilder/MappingBuilder.tsx
Comment (mturley) [inline src/app/Mappings/components/MappingBuilder/MappingBuilder.tsx:36]: We don't need to have a separate state value for whether the button is disabled, since that can be derived from observing the `builderItems`. Keeping it separate and trying to sync the two could cause some bugs.

**Comment by vconzola** [comment]
LGTM. Thanks @gildub .

**Comment by vconzola** [comment]
@gildub Oops! After talking with @mturley I need to rescind my "LGTM" comment. :-(  We want to disable the Add button until a source and target have been selected. Sorry for the confusion.

File: src/app/Mappings/components/MappingBuilder/MappingBuilder.tsx
Comment (mturley) [inline src/app/Mappings/components/MappingBuilder/MappingBuilder.tsx:61]: We should also update our `MOCK_VMWARE_DATASTORES` data so the ids are actually different in each one, which would make this unnecessary-- technically if the items in this list change order or one is removed etc the key is supposed to remain the same for the same object it represents, so React doesn't have to re-render its children. But these components are trivial to render so it's fine

**Comment by gildub** [comment]
@vconzola, that's what I understood and I made the change accordingly just that I hadn't push the update. You can now see it in action.

File: src/app/Mappings/components/AddEditMappingModal.tsx
Comment (mturley) [inline src/app/Mappings/components/AddEditMappingModal.tsx:63]: I don't see what we gain by making this a function, since it is initialized and called all in the same function scope as `builderItems`, and doesn't take any arguments. If we put it after the `useState` below it could just be a variable:

```ts
const isEveryItemFilled = builderItems.every((item) => item.source && item.target);
```

and then similarly the prop type for passing it to MappingBuilder would become a boolean.

I think making things like this a function is only useful if they need to be used in multiple scopes with arguments from each scope, like if we had two different variants of this modal that both had their own set of `builderItem`s.

File: src/app/Mappings/components/MappingBuilder/MappingBuilder.tsx
Comment (mturley) [inline src/app/Mappings/components/MappingBuilder/MappingBuilder.tsx:59]: Same here, I don't think it needs to be a function, just a plain boolean.

File: src/app/Plans/components/Wizard/MappingForm.tsx
Comment (mturley) [inline src/app/Plans/components/Wizard/MappingForm.tsx:65]: Ah, okay, we do use `isEveryItemFilled` in multiple places. I'd say either make both copies of it plain booleans or move it to a `helpers.ts` file somewhere and have it take `builderItems` as an argument. It's pretty trivial, so either way is fine with me, I don't think we lose much having it duplicated.

File: src/app/common/components/AddTooltip.tsx
Comment (mturley) [inline src/app/common/components/AddTooltip.tsx:3]: You should be able to just add TooltipProps to the import from `react-core` in the line above.

**Review by mturley — COMMENTED** [review]
LGTM, just some really minor nitpicks.

File: src/app/common/components/AddTooltip.tsx
Comment (gildub) [reply src/app/common/components/AddTooltip.tsx:3]: I know but it doesn't work for me: 

`Module '"../../../../node_modules/@patternfly/react-core/dist/js"' has no exported member 'TooltipProps'.ts(2305)`

File: src/app/Mappings/components/AddEditMappingModal.tsx
Comment (gildub) [reply src/app/Mappings/components/AddEditMappingModal.tsx:63]: Yes functions are best for reusability but sometimes just for readability.
That said it's called in two different places.

File: src/app/Plans/components/Wizard/MappingForm.tsx
Comment (gildub) [reply src/app/Plans/components/Wizard/MappingForm.tsx:65]: Yes, I agree.

File: src/app/Mappings/components/MappingBuilder/MappingBuilder.tsx
Comment (gildub) [reply src/app/Mappings/components/MappingBuilder/MappingBuilder.tsx]: Ok that one is called only once although I thought this was clearer. I changed it.

**Comment by gildub** [comment]
@vconzola, please review.
One little change: I placed the tooltip on the bottom so it would not hide the selection if it was partly filled.

File: src/app/common/components/AddTooltip.tsx
Comment (mturley) [reply src/app/common/components/AddTooltip.tsx:3]: Aha! Looks like that was a bug in PF, it was supposed to be exported and they fixed it in a newer release. if I update PF on my end it fixes it. Sounds like that's out of scope for this PR though so we can leave this

File: src/app/Mappings/components/AddEditMappingModal.tsx
Comment (mturley) [inline src/app/Mappings/components/AddEditMappingModal.tsx:171]: It doesn't really matter much, but it's still unnecessary for this to be a function. You can just pass `isEveryItemFilled={builderItems.every((item) => item.source && item.target)}` and change the prop to a boolean, and usages from `isEveryItemFilled()` to `isEveryItemFilled`.

File: src/app/Mappings/components/AddEditMappingModal.tsx
Comment (mturley) [inline src/app/Mappings/components/AddEditMappingModal.tsx:98]: Oh, actually @gildub if you're not reusing `isEveryItemFilled` here, you might as well just not even have it be a prop for MappingBuilder, and just have it be defined inside MappingBuilder. It's sort of an implementation detail anyway. I think that might be cleaner... sorry for the nit pick..

File: src/app/Mappings/components/AddEditMappingModal.tsx
Comment (gildub) [reply src/app/Mappings/components/AddEditMappingModal.tsx:171]: Yes, wasn't pay attention there! Thanks

File: src/app/Mappings/components/AddEditMappingModal.tsx
Comment (gildub) [reply src/app/Mappings/components/AddEditMappingModal.tsx:98]: @mturley, sorry I'm not following you here, `isEveryItemFilled` flag needs to be shared by `MappingForm` too.

**Review by mturley — APPROVED** [review]
LGTM, will wait for @vconzola to review before merging.

File: src/app/Mappings/components/AddEditMappingModal.tsx
Comment (mturley) [reply src/app/Mappings/components/AddEditMappingModal.tsx:98]: Oh right, I think I got confused because I saw you wrote out `builderItems.every((item) => item.source && item.target)` twice in this file, so I figured you could just write it out everywhere it's used instead of passing it as a prop. Either way is fine though, it's not worth holding up the PR

Comment (gildub) [reply src/app/Mappings/components/AddEditMappingModal.tsx:98]: Well I see now what you meant.

**Comment by mturley** [comment]
@gildub once you add the tooltip message for the other case (all exhausted) this will be ready to go

**Comment by gildub** [comment]
@mturley, @vconzola, I believe we now have all cases covered.

File: src/app/Mappings/components/MappingBuilder/MappingBuilder.tsx
Comment (mturley) [inline src/app/Mappings/components/MappingBuilder/MappingBuilder.tsx:36]: This creates the messages "all sources network have been mapped" and "all sources storage have been mapped". We need to instead use actual conditions for the two words, so it is phrased "all source networks" and "all source datastores".

**Review by mturley — CHANGES_REQUESTED** [review]

File: src/app/Mappings/components/MappingBuilder/MappingBuilder.tsx
Comment (gildub) [reply src/app/Mappings/components/MappingBuilder/MappingBuilder.tsx]: Done

File: src/app/Mappings/components/MappingBuilder/MappingBuilder.tsx
Comment (mturley) [inline src/app/Mappings/components/MappingBuilder/MappingBuilder.tsx:36]: @gildub this is still going to produce the wrong spelling/phrasing.
1. `mappingType` is either Network or Storage, we need "networks" and *"datastores"*
2. The word "sources" should not be plural, only the type.

We need this:
```ts
  const messageExhausted = `All source ${
    mappingType === MappingType.Network ? 'networks' : 'datastores'
  } have been mapped.`;
```

Comment (gildub) [reply src/app/Mappings/components/MappingBuilder/MappingBuilder.tsx:36]: Sorry, I shouldn't have been working on it late last night, not paying enough attention.

Comment (mturley) [reply src/app/Mappings/components/MappingBuilder/MappingBuilder.tsx]: No problem at all!

Yeah, it's JS expressions all the way down :)

**Review by mturley — APPROVED** [review]
Looks great!

**Pull request merged**
